src/codegen.cpp

```diff
@@ -9123,21 +9118,18 @@ void add_cc_args(CodeGen *g, ZigList<const char *> &args, const char *out_dep_pa
         args.append("-target");
         args.append(buf_ptr(&g->llvm_triple_str));
 
-        if (target_is_musl(g->zig_target) && target_is_riscv(g->zig_target)) {
-            // Musl depends on atomic instructions, which are disabled by default in Clang/LLVM's
-            // cross compilation CPU info for RISCV.
-            // TODO: https://github.com/ziglang/zig/issues/2883
+        if (g->target_details) {
             args.append("-Xclang");
-            args.append("-target-feature");
+            args.append("-target-cpu");
             args.append("-Xclang");
-            args.append(riscv_default_features);
-        } else if (g->zig_target->os == OsFreestanding && g->zig_target->arch == ZigLLVM_x86) {
+            args.append(stage2_target_details_get_llvm_cpu(g->target_details));
```

Comment:
is this correct if the cpu is null?

Reply:
the `get_llvm_{cpu,features}` functions always return a valid string. If features were provided, an empty string is then passed as the CPU, which LLVM interprets as "the user has not chosen a CPU". When specific features are provided, the llvm CPU string is set to the empty string [here](https://github.com/layneson/zig/blob/4e0e004bb91851d43c8f0a3c93e4b2cb2adde232/src-self-hosted/stage1.zig#L703).